From: ArmOS Project <devnull@armos.local>
Subject: [PATCH] gallium: add the ArmOS VirGL winsys and frontend hooks

ArmOS links Mesa statically and exposes VirGL through its own portable DRM
userspace ABI.  It therefore has neither libdl nor the Linux DRM/vtest
winsyses.  Select the ArmOS adapter explicitly without changing the other
platform paths.  Build the architecture-neutral ArmGL state-tracker frontend
separately from the selected GPU winsys.
---
 meson.build                              | 6 +++++-
 meson.options                            | 7 +++++++
 src/gallium/drivers/virgl/meson.build   | 4 +++-
 src/gallium/drivers/virgl/virgl_context.c | 2 +-
 src/gallium/meson.build                 | 7 +++++--
 5 files changed, 21 insertions(+), 5 deletions(-)

diff --git a/meson.build b/meson.build
index 1111111..2222222 100644
--- a/meson.build
+++ b/meson.build
@@ -1072,6 +1072,8 @@ foreach a : ['warn_unused_result', 'weak', 'format', 'packed', 'returns_nonnull',
 # TODO: this is very incomplete
 if ['linux', 'cygwin', 'gnu', 'freebsd', 'gnu/kfreebsd', 'haiku', 'android', 'managarm'].contains(host_machine.system())
   pre_args += '-D_GNU_SOURCE'
+elif host_machine.system() == 'armos'
+  pre_args += '-D_POSIX_C_SOURCE=200809L'
 elif host_machine.system() == 'sunos'
   pre_args += '-D__EXTENSIONS__'
 elif host_machine.system() == 'windows'
@@ -1535,7 +1535,7 @@ ld_args_build_id = cc.get_supported_link_arguments('-Wl,--build-id=sha1')
 
 # check for dl support
 dep_dl = null_dep
-if host_machine.system() != 'windows'
+if not ['windows', 'armos'].contains(host_machine.system())
   if not cc.has_function('dlopen')
     dep_dl = cc.find_library('dl', required : true)
   endif
@@ -1594,9 +1594,12 @@ endif
 
 if host_machine.system() == 'windows'
   # For MSVC and MinGW we aren't using pthreads, and dependency('threads') will add linkage
   # to pthread for MinGW, so leave the dependency null_dep for Windows. For Windows linking to
   # kernel32 is enough for c11/threads.h and it's already linked by meson by default
   dep_thread = null_dep
+elif host_machine.system() == 'armos'
+  # ArmOS pthread symbols are provided by the static userland runtime.
+  dep_thread = declare_dependency()
 else
   dep_thread = dependency('threads')
 endif
diff --git a/src/gallium/drivers/virgl/virgl_context.c b/src/gallium/drivers/virgl/virgl_context.c
index 9999999..aaaaaaa 100644
--- a/src/gallium/drivers/virgl/virgl_context.c
+++ b/src/gallium/drivers/virgl/virgl_context.c
@@ -22,7 +22,7 @@
  */
 
 #include <string.h>
-#ifndef _WIN32
+#if !defined(_WIN32) && !defined(__ARMOS__)
 #include <libsync.h>
 #endif
 
diff --git a/src/gallium/meson.build b/src/gallium/meson.build
index 3333333..4444444 100644
--- a/src/gallium/meson.build
+++ b/src/gallium/meson.build
@@ -170,7 +173,9 @@ endif
 endif
 if with_gallium_virgl
   subdir('winsys/virgl/common')
-  if not with_platform_windows
+  if host_machine.system() == 'armos'
+    subdir('winsys/virgl/armos')
+  elif not with_platform_windows
     subdir('winsys/virgl/drm')
     subdir('winsys/virgl/vtest')
   endif
@@ -230,6 +233,9 @@ if with_dri
 endif
 if with_platform_haiku
   subdir('frontends/hgl')
 endif
+if host_machine.system() == 'armos'
+  subdir('frontends/armgl')
+endif
 if with_gallium_st_d3d10umd
   subdir('frontends/d3d10umd')
diff --git a/meson.options b/meson.options
index 7777777..8888888 100644
--- a/meson.options
+++ b/meson.options
@@ -1,6 +1,13 @@
 # Copyright © 2017-2019 Intel Corporation
 # SPDX-License-Identifier: MIT
 
+option(
+  'armos-virgl-winsys-dir',
+  type : 'string',
+  value : '',
+  description : 'Target directory containing libarmos-virgl-winsys.a',
+)
+
 option(
   'split-debug',
   type : 'feature',
diff --git a/src/gallium/drivers/virgl/meson.build b/src/gallium/drivers/virgl/meson.build
index 5555555..6666666 100644
--- a/src/gallium/drivers/virgl/meson.build
+++ b/src/gallium/drivers/virgl/meson.build
@@ -25,7 +25,9 @@ libvirgl = static_library(
 )
 
 virgl_deps = [libvirgl]
-if not with_platform_windows
+if host_machine.system() == 'armos'
+  virgl_deps += libvirglarmos
+elif not with_platform_windows
   virgl_deps += libvirgldrm
   virgl_deps += libvirglvtest
 endif
